### PR TITLE
Refactor java_bytecode_convert_methodt::convert_instructions

### PR DESCRIPTION
**Pull request by romainbrenguier**
The goal of this PR is to reduce the size of this huge method without changing its behaviour, mainly by extracting some bits into new methods.

### CODE REVIEW

File: src/util/invariant.h [hunk 1]
```diff
@@ -202,19 +202,21 @@ inline void invariant_violated_string(
...
-  if(!(CONDITION)) \
-    invariant_violated_string(__FILE__, __this_function__, __LINE__, (REASON)); /* NOLINT */  \
-  } while(0)
+#define INVARIANT(CONDITION, REASON)                                           \
+  do /* NOLINT */                                                              \
+  {                                                                            \
+    if(!(CONDITION))                                                           \
+      invariant_violated_string(                                               \
+        __FILE__, __this_function__, __LINE__, (REASON)); /* NOLINT */         \
+  } while(false)
```
Comment (tautschnig) [inline src/util/invariant.h]: 1) Which are these "some" warnings? 2) The previous code even worked within C code, while this is now C++; is that really necessary? 3) Please place this commit in a separate PR.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 2]
```diff
@@ -1123,28 +1093,23 @@ codet java_bytecode_convert_methodt::convert_instructions(
 
   while(!working_set.empty())
   {
-    std::set<unsigned>::iterator cur=working_set.begin();
-    address_mapt::iterator a_it=address_map.find(*cur);
-    CHECK_RETURN(a_it != address_map.end());
+    auto cur = working_set.begin();
+    auto a = address_map.at(*cur);
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: I don't think this is semantics preserving: the previous code would result in modifications to the mapped element, while all those now happen on a copy. `auto` is dangerous.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -975,10 +975,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
   std::vector<unsigned> jsr_ret_targets;
   std::vector<instructionst::const_iterator> ret_instructions;
 
-  for(instructionst::const_iterator
-      i_it=instructions.begin();
-      i_it!=instructions.end();
-      i_it++)
+  for(auto i_it = instructions.begin(); i_it != instructions.end(); i_it++)
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: I'm not sure I see much value in this change.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 4]
```diff
@@ -2711,6 +2676,45 @@ codet java_bytecode_convert_methodt::convert_instructions(
   return code;
 }
 
+void java_bytecode_convert_methodt::draw_edges_from_ret_to_jsr(
+  java_bytecode_convert_methodt::address_mapt &address_map,
+  std::vector<unsigned int> &jsr_ret_targets,
+  const std::vector<
+    std::vector<java_bytecode_parse_treet::instructiont>::const_iterator>
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Please introduce a typedef, this is very hard to parse for humans, and the same type is used multiple times.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 5]
```diff
@@ -2711,6 +2676,45 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+}
+
+void java_bytecode_convert_methodt::add_try_catch_handler_to_target(
+  const java_bytecode_convert_methodt::methodt &method,
+  std::set<unsigned int> &targets,
+  const std::vector<java_bytecode_parse_treet::instructiont>::const_iterator
+    &i_it,
+  const std::pair<
+    std::map<unsigned int,
+             java_bytecode_convert_methodt::converted_instructiont>::iterator,
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: You only ever use `a_entry.first->second`, so pass some simpler member.

**Review by tautschnig — COMMENTED** [review]
The last two commits are rather impossible to review for me at the moment (requires a big screen); detailed comments for the first two only.

File: src/util/invariant.h (hunk 1 above)
Comment (romainbrenguier) [reply src/util/invariant.h]: 1. CLion warnings when INVARIANT is used 2. `false` is already used in this file, so I don't think this changes anything. 3. done here: https://github.com/diffblue/cbmc/pull/2078

**Comment by romainbrenguier** [comment]
@tautschnig 
> The last two commits are rather impossible to review for me at the moment (requires a big screen); detailed comments for the first two only.

I currently trying to split these commits, so if you wait a bit, it should become easier to review

**Comment by peterschrammel** [comment]
@romainbrenguier, what's the status of this?

**Comment by romainbrenguier** [comment]
@peterschrammel I just rebased

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 6]
```diff
@@ -1065,10 +1062,10 @@ codet java_bytecode_convert_methodt::convert_instructions(
       if(i_it->statement=="jsr" ||
          i_it->statement=="jsr_w")
       {
-        instructionst::const_iterator next=i_it+1;
         assert(
           next!=instructions.end() &&
           "jsr without valid return address?");
+        auto next = i_it + 1;
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: FYI: You could replace that with `std::next(i_it)`.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 7]
```diff
@@ -2313,6 +2289,38 @@ codet java_bytecode_convert_methodt::convert_instructions(
   return code;
 }
 
+void java_bytecode_convert_methodt::convert_multianewarray(
+  const std::vector<java_bytecode_parse_treet::instructiont>::const_iterator
+    &i_it,
+  exprt &arg0,
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Make parameters `const` where possible

File: jbmc/src/java_bytecode/java_bytecode_convert_method_class.h [hunk 8]
```diff
@@ -414,5 +414,13 @@ class java_bytecode_convert_methodt:public messaget
...
     codet &c,
     exprt::operandst &results);
+
+  void convert_multianewarray(
+    const std::vector<java_bytecode_parse_treet::instructiont>::const_iterator
+      &i_it,
+    exprt &arg0,
+    exprt::operandst &op,
+    codet &c,
+    exprt::operandst &results);
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_convert_method_class.h]: Make methods `const` where possible.

**Review by peterschrammel — COMMENTED** [review]
Thanks @romainbrenguier for this extensive piece of work. Please check all methods that you have added whether you can make them or some of their parameters `const`.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 9]
```diff
@@ -2720,6 +2710,22 @@ codet java_bytecode_convert_methodt::convert_instructions(
   return code;
 }
 
+void java_bytecode_convert_methodt::add_try_catch_handler_to_target(
+  const java_bytecode_convert_methodt::methodt &method,
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Passing the entire method might not be necessary, you're just using the `exception_table` in the function body.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 10]
```diff
@@ -2720,6 +2710,22 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+  std::set<unsigned int> &targets,
+  const unsigned int address,
+  java_bytecode_convert_methodt::converted_instructiont &instruction) const
+{
+  for(const auto &exception_row : method.exception_table)
+  {
+    if(address >= exception_row.start_pc && address < exception_row.end_pc)
+    {
+      instruction.successors.push_back(exception_row.handler_pc);
+      targets.insert(exception_row.handler_pc);
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Thought: this method could've returned a vector of unsigned ints, rather than storing results to two seemingly unrelated containers. Those containers only make sense from the point of view of the outer algorithm, but not necessarily from here.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 11]
```diff
@@ -2710,6 +2700,22 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
 
+void java_bytecode_convert_methodt::draw_edges_from_ret_to_jsr(
+  java_bytecode_convert_methodt::address_mapt &address_map,
+  std::vector<unsigned int> &jsr_ret_targets,
+  const std::vector<
+    std::vector<java_bytecode_parse_treet::instructiont>::const_iterator>
+    &ret_instructions) const
+{ // Draw edges from every `ret` to every `jsr` successor. Could do better with
+  // flow analysis to distinguish multiple subroutines within the same function.
+  for(const auto retinst : ret_instructions)
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Unnecessary copying. Use `const auto&`.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 12]
```diff
@@ -1112,28 +1107,23 @@ codet java_bytecode_convert_methodt::convert_instructions(
 
   while(!working_set.empty())
   {
-    address_mapt::iterator a_it=address_map.find(*cur);
-    CHECK_RETURN(a_it != address_map.end());
     auto cur = working_set.begin();
+    auto &a = address_map.at(*cur);
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: A more descriptive name would be appreciated.

Also, keep in mind, that `.at` _throws_ on failure. If you want this to behave like an invariant, you have to mark this function `noexcept`. And the thing to keep in mind about `noexcept` it doesn't work in Visual Studio 2013 IIRC, so you'll have to resort to some kind of macro.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 13]
```diff
@@ -2690,6 +2663,39 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+
+  pop(parameters.size());
+
+  const typet &return_type = code_type.return_type();
+
+  if(return_type.id() != ID_empty)
+  {
+    results.resize(1);
+    results[0] = zero_initializer(
+      return_type, location, namespacet(symbol_table), get_message_handler());
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: This looks like an ideal candidate for an `optional` return.

**Review by LAJW — APPROVED** [review]
Good extracting, the only complaint I have is that it's a little bit too greedy - functions you've extracted take in too much state. Rather than passing in an `[out]` parameter(s), you should return the value. Many times you're using only a single field/object from the object being passed in. The function becomes unnecessarily complex, and if anyone were to write unit tests for it, he'd have to create a huge object, where only a single sub element of it is tested and actually matters. Also variable names could be expanded `c`, `a`, `arg0`.

**Comment by romainbrenguier** [comment]
@peterschrammel I added as much const as I could
@LAJW yes that was pretty greedy, I got rid of the iterator arguments that were used for just one element, that should be better

**Review by peterschrammel — APPROVED** [review]
LGTM modulo TG bump passing
